File: sdks/android/streamkit-sdk/src/main/kotlin/com/rajutechie/streamkit/StreamKit.kt
package com.rajutechie.streamkit

import android.content.Context
import com.rajutechie.streamkit.modules.ChatModule
import com.rajutechie.streamkit.modules.CallModule
import com.rajutechie.streamkit.modules.MeetingModule
import com.rajutechie.streamkit.modules.StreamModule
import com.rajutechie.streamkit.transport.HttpClient
import com.rajutechie.streamkit.transport.WebSocketClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob

data class RajutechieStreamKitConfig(
    val apiKey: String,
    val region: Region = Region.US_EAST_1,
    val apiUrl: String = "https://api.rajutechie-streamkit.io/v1",
    val wsUrl: String = "wss://ws.rajutechie-streamkit.io",
    val logLevel: LogLevel = LogLevel.WARN
)

enum class Region { US_EAST_1, US_WEST_2, EU_WEST_1, AP_SOUTHEAST_1 }
enum class LogLevel { DEBUG, INFO, WARN, ERROR }

class RajutechieStreamKit private constructor(
    private val context: Context,
    private val config: RajutechieStreamKitConfig
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val httpClient = HttpClient(config.apiUrl, config.apiKey)
    private val wsClient = WebSocketClient(config.wsUrl)

    val chat: ChatModule by lazy { ChatModule(httpClient, wsClient, scope) }
    val call: CallModule by lazy { CallModule(httpClient, wsClient, scope) }
    val meeting: MeetingModule by lazy { MeetingModule(httpClient, wsClient, scope) }
    val stream: StreamModule by lazy { StreamModule(httpClient, wsClient, scope) }

    var isConnected: Boolean = false
        private set

    suspend fun connect(userToken: String) {
        httpClient.setToken(userToken)
        wsClient.connect(userToken)
        isConnected = true
    }

    suspend fun disconnect() {
        wsClient.disconnect()
        httpClient.setToken(null)
        isConnected = false
    }

    companion object {
        @Volatile
        private var instance: RajutechieStreamKit? = null

        fun initialize(context: Context, config: RajutechieStreamKitConfig): RajutechieStreamKit {
            return instance ?: synchronized(this) {
                instance ?: RajutechieStreamKit(context.applicationContext, config).also { instance = it }
            }
        }

        fun getInstance(): RajutechieStreamKit {
            return instance ?: throw IllegalStateException("RajutechieStreamKit not initialized. Call initialize() first.")
        }
    }
}

File: sdks/ios/Sources/RajutechieStreamKit/StreamKit.swift
import Foundation

public struct RajutechieStreamKitConfig {
    public let apiKey: String
    public let region: Region
    public let apiUrl: String
    public let wsUrl: String

    public init(
        apiKey: String,
        region: Region = .usEast1,
        apiUrl: String = "https://api.rajutechie-streamkit.io/v1",
        wsUrl: String = "wss://ws.rajutechie-streamkit.io"
    ) {
        self.apiKey = apiKey
        self.region = region
        self.apiUrl = apiUrl
        self.wsUrl = wsUrl
    }
}

public enum Region: String {
    case usEast1 = "us-east-1"
    case usWest2 = "us-west-2"
    case euWest1 = "eu-west-1"
    case apSoutheast1 = "ap-southeast-1"
}

public enum ConnectionState {
    case connecting, connected, disconnected, reconnecting
}

@MainActor
public final class RajutechieStreamKit {
    public static let shared = RajutechieStreamKit()

    private var config: RajutechieStreamKitConfig?
    private var httpClient: HTTPClient?
    private var wsClient: WebSocketClient?

    public private(set) var connectionState: ConnectionState = .disconnected

    private init() {}

    public func configure(_ config: RajutechieStreamKitConfig) -> RajutechieStreamKit {
        self.config = config
        self.httpClient = HTTPClient(baseURL: config.apiUrl, apiKey: config.apiKey)
        self.wsClient = WebSocketClient(url: config.wsUrl)
        return self
    }

    public var chat: ChatModule {
        guard let http = httpClient, let ws = wsClient else {
            fatalError("RajutechieStreamKit not configured. Call configure() first.")
        }
        return ChatModule(http: http, ws: ws)
    }

    public var call: CallModule {
        guard let http = httpClient, let ws = wsClient else {
            fatalError("RajutechieStreamKit not configured. Call configure() first.")
        }
        return CallModule(http: http, ws: ws)
    }

    public var meeting: MeetingModule {
        guard let http = httpClient, let ws = wsClient else {
            fatalError("RajutechieStreamKit not configured. Call configure() first.")
        }
        return MeetingModule(http: http, ws: ws)
    }

    public var stream: StreamModule {
        guard let http = httpClient, let ws = wsClient else {
            fatalError("RajutechieStreamKit not configured. Call configure() first.")
        }
        return StreamModule(http: http, ws: ws)
    }

    public func connect(token: String) async throws {
        connectionState = .connecting
        await httpClient?.setToken(token)
        try await wsClient?.connect(token: token)
        connectionState = .connected
    }

    public func disconnect() async {
        await wsClient?.disconnect()
        await httpClient?.setToken(nil)
        connectionState = .disconnected
    }
}
